### PR TITLE
Fix validation skipped when IterableDataset exhausts before reported length

### LINKED ISSUE #19624: IterableDataset with CORRECT length causes validation loop to be skipped
### Bug description

This is related to this issue:
https://github.com/Lightning-AI/pytorch-lightning/issues/10290

Whereby an IterableDataset with a length defined wont trigger a validation epoch, even if the defined length is correct so long as the following conditions met:

1. Accurate length of IterableDataset defined
2. Dataset accurately split between multiple workers with no overlap
3. Drop last = True for the dataloader
4. Dataset size does not evenly divide into the batches

In this instance multiple workers may be left with an incomplete batch right at the end of the training epoch. So the number of "dropped batches" exceeds 1. Then the dataloader will raise a StopIteration before the length is reached, causing the validation epoch to be skipped.

This is standard PyTorch behavior as the collation function is called per worker in an IterableDataset.
https://github.com/pytorch/pytorch/issues/33413

I am having this issue right now my current fix is artificially subtract from the length of my IterableDataset to account for this. Unfortunately I really would like the length to be defined, so can't set it to inf which was the hotfix in the previous thread. 
The progress bar is useful for me to judge which partition I need to run certain jobs on plus I use the dataset length to sync up my cyclic learning rate with the number of steps in an epoch. 

### What version are you seeing the problem on?

master

### How to reproduce the bug

```python
import torch as T
import numpy as np
from torch.utils.data import DataLoader, Dataset, IterableDataset, get_worker_info

from lightning import LightningModule, Trainer, LightningDataModule

nwrkrs = 4
drop = True

class Data(IterableDataset):
    def __init__(self) -> None:
        super().__init__()
        self.data = np.random.rand(100, 10).astype(np.float32)

    def __len__(self) -> int:
        return len(self.data)

    def __iter__(self):
        worker_info = get_worker_info()
        worker_id = 0 if worker_info is None else worker_info.id
        num_workers = 1 if worker_info is None else worker_info.num_workers
        worker_samples = np.array_split(self.data, num_workers)[worker_id]

        for i in worker_samples:
            yield i

class Model(LightningModule):
    def __init__(self) -> None:
        super().__init__()
        self.layer = T.nn.Linear(10, 1)
        self.did_validation = False

    def forward(self, x: T.Tensor) -> T.Tensor:
        return self.layer(x)

    def training_step(self, batch):
        return self(batch).mean()

    def validation_step(self, batch):
        self.did_validation = True
        return self(batch).mean()

    def configure_optimizers(self):
        return T.optim.Adam(self.parameters())

model = Model()
trainer = Trainer(logger=False, max_epochs=2, num_sanity_val_steps=0)
train_loader = DataLoader(Data(), num_workers=nwrkrs, batch_size=32, drop_last=drop)
valid_loader = DataLoader(Data(), num_workers=nwrkrs, batch_size=32, drop_last=drop)
trainer.fit(model, train_loader, valid_loader)
print("Performed validation:", model.did_validation)
```

Setting up the code above and running it with the following settings gives these results:

```python
nwrkrs = 0, drop = True
```
`Performed validation: True`
```python
nwrkrs = 4, drop = False
```
`Performed validation: True`
```python
nwrkrs = 4, drop = True
```
`Performed validation: False`

cc @justusschock @awaelchli

### PR DESCRIPTION
**Pull request by avocardio**
## Bug

Fixes #19624

When an `IterableDataset` reports a length via `__len__` but produces fewer batches than expected (e.g. due to `drop_last=True` with multiple workers, shard boundary rounding in webdataset/DALI, or worker-based splitting), `StopIteration` is raised before the expected batch count. This causes `on_advance_end` to be skipped for the final iteration, which means the end-of-epoch validation check **never runs** — not just for that epoch, but for all subsequent epochs as well.

### Root cause

In `_TrainingEpochLoop.run()`:

```python
while not self.done:
    try:
        self.advance(data_fetcher)
        self.on_advance_end(data_fetcher)
    except StopIteration:
        break
```

When `_DataFetcher.__next__()` encounters a `StopIteration` from the underlying iterator, it sets `self.done = True` and then **re-raises** the exception. The `except StopIteration: break` in the training loop catches this, which skips `on_advance_end()` — where the validation check (`_should_check_val_fx`) lives.

### Fix

After the while loop exits, check whether the data fetcher was exhausted (`data_fetcher.done`) and if so, run the end-of-epoch validation check. This only triggers validation — not the per-batch operations (LR scheduler updates, `_batches_that_stepped` increment, logger saves) that correctly should not run when no batch was processed.

### How to reproduce

```python
from torch.utils.data import DataLoader, IterableDataset

class ShortIterableDataset(IterableDataset):
    """Reports length=10 but only yields 8 samples."""
    def __iter__(self):
        for i in range(8):
            yield torch.randn(32)
    def __len__(self):
        return 10

trainer = Trainer(max_epochs=2, num_sanity_val_steps=0)
trainer.fit(model, DataLoader(ShortIterableDataset(), batch_size=2))
# Validation never runs (0 times instead of 2)
```

This is common with webdataset/DALI streaming datasets where `__len__` is an estimate based on total samples / batch_size but the actual number of batches produced can vary due to shard boundaries, worker splitting, or `drop_last`.

## Test plan

- Added `test_iterable_dataset_validation_on_exhaustion` regression test that creates a `ShortIterableDataset` reporting `__len__() = 10` but yielding only 8 samples (4 batches vs expected 5), and verifies validation runs for both epochs
- Verified fix in production: two 40-epoch multi-GPU training runs on GH200 480GB servers with webdataset IterableDatasets, confirming validation fires reliably every epoch after applying this fix (previously validation was permanently skipped after epoch 0)

----
📚 Documentation preview 📚: https://pytorch-lightning--21551.org.readthedocs.build/en/21551/

### CODE REVIEW

**Comment by avocardio** [comment]
Closing to resubmit with a cleaner description.
